### PR TITLE
[RemoteAST/Reflection] Changes to MetadataReader and its users to track function parameter flags

### PR DESCRIPTION
**Pull request by xedin**

### CODE REVIEW

File: include/swift/Remote/MetadataReader.h [hunk 1]
```diff
@@ -29,6 +29,42 @@
 namespace swift {
 namespace remote {
 
+template <typename BuiltType> class FunctionParam {
+  BuiltType Type;
+  StringRef Label;
+
+  union {
```
Comment (DougGregor) [inline include/swift/Remote/MetadataReader.h]: This shouldn't be a union.

File: include/swift/Remote/MetadataReader.h [hunk 2]
```diff
@@ -745,15 +813,20 @@ class MetadataReader {
...
         const auto InOutMask = (StoredPointer) 1;
-        ArgumentIsInOut.push_back((FlaggedArgumentAddress & InOutMask) != 0);
-        FlaggedArgumentAddress &= ~InOutMask;
+        if ((FlaggedArgumentAddress & InOutMask) != 0)
+          Param.setInOut();
 
-        if (auto ArgumentTypeRef = readTypeFromMetadata(FlaggedArgumentAddress))
-          Arguments.push_back(ArgumentTypeRef);
-        else
+        FlaggedArgumentAddress &= ~InOutMask;
```
Comment (DougGregor) [inline include/swift/Remote/MetadataReader.h]: `shared` and `variadic`, too? It's odd to have them partially-supported, so I suggest either completing support or leaving a FIXME to do it all later.

Comment (xedin) [reply include/swift/Remote/MetadataReader.h]: Yes, that's the next step after `MetadataReader` is updated to support these flags I'm going to update function metadata itself to include all of the flags for parameters...

File: include/swift/Remote/MetadataReader.h (hunk 1 above)
Comment (xedin) [reply include/swift/Remote/MetadataReader.h]: Ah, sorry, I wanted to change that to be to enum like other flags we have...

File: include/swift/Remote/MetadataReader.h [hunk 3]
```diff
@@ -29,6 +30,28 @@
 namespace swift {
 namespace remote {
 
+template <typename BuiltType> class FunctionParam {
+  BuiltType Type;
+  StringRef Label;
+  ParameterTypeFlags Flags;
```
Comment (xedin) [inline include/swift/Remote/MetadataReader.h]: @DougGregor please let me know what you think about this? I'm not sure if it's comme il faut to use structures from `AST/Types.h` in here.

File: include/swift/Remote/MetadataReader.h (hunk 2 above)
Comment (xedin) [reply include/swift/Remote/MetadataReader.h]: Added a FIXME about it on L807, going to address after this changes are in.

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please clean test

**Comment by xedin** [comment]
@slavapestov it looks like I should be bugging you about changes to Reflection/RemoteAST instead :)

**Comment by xedin** [comment]
@swift-ci please clean test

**Pull request merged**